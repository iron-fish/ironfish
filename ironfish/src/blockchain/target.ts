/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

import type Serde from '../serde'

function max(a: BigInt, b: BigInt): BigInt {
  if (a > b) {
    return a
  } else {
    return b
  }
}

/**
 * Courtesy of https://coolaj86.com/articles/convert-js-bigints-to-typedarrays/
 *
 * Convert a Buffer to a big integer number, in big endian format.
 *
 * I'm concerned about efficiency here. Converting a string and back and... WTF?
 * Every block hash attempt has to be converted to a Target, so this is a function
 * that should be optimized. We may want to compile this to wasm if there isn't
 * a less janky way to do it.
 *
 * I'm pushing it out like this for now so I can focus on bigger architecture concerns.
 *
 * Sorry.
 */
export function bytesToBigInt(bytes: Buffer): bigint {
  const hex: string[] = []
  if (bytes.length === 0) {
    return BigInt(0)
  }
  bytes.forEach(function (i) {
    let h = i.toString(16)
    if (h.length % 2) {
      h = '0' + h
    }
    hex.push(h)
  })

  return BigInt('0x' + hex.join(''))
}

export function bigIntToBytes(bigint: bigint): Buffer {
  let hex = bigint.toString(16)
  if (hex.length % 2) {
    hex = '0' + hex
  }

  const len = hex.length / 2
  const u8 = Buffer.alloc(len)

  let i = 0
  let j = 0
  while (i < len) {
    u8[i] = parseInt(hex.slice(j, j + 2), 16)
    i += 1
    j += 2
  }

  return u8
}

/**
 * The bound divisor of the difficulty, used to update difficulty (and subsequently target).
 * We are taking in large part Ethereum's dynamic difficulty calculation,
 * with the exeption of 'uncles' and 'difficulty bomb' as a concept
 * https://github.com/ethereum/EIPs/blob/master/EIPS/eip-2.md
 * original algorithm:
 * diff = (parent_diff +
 *         (parent_diff / 2048 * max(1 - (current_block_timestamp - parent_timestamp) // 10, -99))
 *        ) + 2**((current_block_number // 100000) — 2)
 * Note we are not including the difficulty bomb (which is this part: 2**((current_block_number // 100000) — 2))
 * So the algorithm for target is:
 * diff = parent_diff + parent_diff / 2048 * max(1 - (current_block_timestamp - parent_timestamp) / 10, -99)
 * note that timestamps above are in seconds, and JS timestamps are in ms
 * The bound divisor of the difficulty is the '2048' part of that equation
 */
const DIFFICULTY_ADJUSTMENT_DENOMINATOR = 2048

/**
 *  Minimum difficulty, which is equivalent to maximum target
 */
const MIN_DIFFICULTY = BigInt(131072)

/**
 * Maximum target, which is equivalent of minimum difficulty of 131072
 * target == 2**256 / difficulty
 */
const MAX_TARGET = BigInt(
  '883423532389192164791648750371459257913741948437809479060803100646309888',
)

/**
 *  Maximum number to represent a 256 bit number, which is 2**256 - 1
 */
const MAX_256_BIT_NUM = BigInt(
  '115792089237316195423570985008687907853269984665640564039457584007913129639935',
)

export class Target {
  targetValue: bigint
  constructor(targetValue: bigint | BigInt | Buffer | string | number | undefined = undefined) {
    if (targetValue === undefined) {
      this.targetValue = BigInt(0)
    } else {
      let candidate
      if (targetValue instanceof Buffer) {
        candidate = bytesToBigInt(targetValue)
      } else {
        candidate = BigInt(targetValue)
      }
      if (candidate > MAX_256_BIT_NUM) {
        throw new Error('Target value exceeds max target')
      } else {
        this.targetValue = candidate
      }
    }
  }

  /**
   * Maximum target (in terms of difficulty), which is equivalent of
   * minimum difficulty of 131072
   * maximum target == minimum difficulty
   * target == 2**256 / difficulty
   */
  static maxTarget(): Target {
    return new Target(MAX_TARGET)
  }

  /**
   * The initial target on the genesis block.
   *
   * This will need to be hand-tuned.
   */
  static initialTarget(): Target {
    return this.maxTarget()
  }

  /**
   * Calculate the target for the current block given the timestamp in that
   * block's header, the pervious block's timestamp and previous block's target.
   *
   * To verify whether a target on a block is correct, pass in the timestamp in its header,
   * its previous block's timestamp, and its previous block's target
   * and compare the resulting target to what is specified on the current block header
   *
   * @param time the block's timestamp for which the target is calculated for
   * @param previousBlockTimestamp the block's previous block header's timestamp
   * @param previousBlockTarget the block's previous block header's target
   */
  static calculateTarget(
    time: Date,
    previousBlockTimestamp: Date,
    previousBlockTarget: Target,
  ): Target {
    const parentDifficulty = previousBlockTarget.toDifficulty()
    const difficulty = Target.calculateDifficulty(
      time,
      previousBlockTimestamp,
      parentDifficulty,
    )

    return Target.fromDifficulty(difficulty)
  }

  /**
   *
   * Calculate the difficulty for the current block given the timestamp in that
   * block's header, the pervious block's timestamp and previous block's target.
   *
   * Note that difficulty == 2**256 / target and target == 2**256 / difficulty
   *
   * Returns the difficulty for a block given it timestamp for that block and its parent.
   * @param time the block's timestamp for which the target is calcualted for
   * @param previousBlockTimestamp the block's previous block header's timestamp
   * @param previousBlockTarget the block's previous block header's target
   */
  static calculateDifficulty(
    time: Date,
    previousBlockTimestamp: Date,
    previousBlockDifficulty: BigInt,
  ): BigInt {
    // We are taking in large part Ethereum's dynamic difficulty calculation,
    // with the exeption of 'uncles' and 'difficulty bomb' as a concept
    // https://github.com/ethereum/EIPs/blob/master/EIPS/eip-2.md
    // original algorithm:
    // diff = (parent_diff +
    //         (parent_diff / 2048 * max(1 - (current_block_timestamp - parent_timestamp) // 10, -99))
    //        ) + 2**((current_block_number // 100000) — 2)
    // Note we are not including the difficulty bomb (which is this part: 2**((current_block_number // 100000) — 2))
    // So the algorithm we're taking is:
    // diff = parent_diff + parent_diff / 2048 * max(1 - (current_block_timestamp - parent_timestamp) / 10, -99)
    // note that timestamps above are in seconds, and JS timestamps are in ms

    // max(1 - (current_block_timestamp - parent_timestamp) / 10, -99)
    const diffInSeconds = (time.getTime() - previousBlockTimestamp.getTime()) / 1000
    const sign = BigInt(Math.max(1 - Math.floor(diffInSeconds / 10), -99))
    const offset = BigInt(previousBlockDifficulty) / BigInt(DIFFICULTY_ADJUSTMENT_DENOMINATOR)

    // diff = parent_diff + parent_diff / 2048 * max(1 - (current_block_timestamp - parent_timestamp) / 10, -99)
    const difficulty = max(
      BigInt(previousBlockDifficulty) + offset * sign,
      Target.minDifficulty(),
    )

    return difficulty
  }

  /**
   * Returns the minimum difficulty that can be used for Iron Fish blocks
   * To be used in calculateTarget for easier mocking
   */
  static minDifficulty(): BigInt {
    return MIN_DIFFICULTY
  }

  /**
   * Converts difficulty to Target
   */
  static fromDifficulty(difficulty: BigInt): Target {
    if (difficulty == BigInt(1)) {
      return new Target(MAX_256_BIT_NUM)
    }
    return new Target(BigInt(2) ** BigInt(256) / BigInt(difficulty))
  }

  /**
   * Return the difficulty representation as a big integer
   */
  toDifficulty(): bigint {
    return BigInt(2) ** BigInt(256) / this.targetValue
  }

  /**
   * Add the given amout to the target's value. A negative amount makes the target
   * harder to achieve, a positive one makes it easier.
   *
   * If adjustment would make target negative or higher than max allowed value,
   * the current target is returned unchanged.
   */
  adjust(amount: bigint): Target {
    let adjusted = this.targetValue + amount
    if (adjusted > MAX_TARGET || adjusted < 0) {
      adjusted = this.targetValue
    }
    return new Target(adjusted)
  }

  /**
   * Return whether or not this target meets the requirements of the given target,
   * which is to say, this has a lower numeric value then the provided one.
   */
  static meets(hashValue: BigInt, target: Target): boolean {
    return hashValue <= target.targetValue
  }

  /**
   * Return the target number as a big integer
   */
  asBigInt(): bigint {
    return this.targetValue
  }

  /**
   * Big endian, and since even after 20 years in the industry, I can't remember
   * what that means, the most significant byte is in the 0th index of the array.
   *
   * The resulting byte array is always left padded with 0s to be 32 bytes long
   */
  asBytes(): Buffer {
    const bytes = bigIntToBytes(this.targetValue)
    const result = Buffer.alloc(32)
    result.set(bytes, 32 - bytes.length)
    return result
  }
}

export class TargetSerde implements Serde<Target, string> {
  equals(target1: Target, target2: Target): boolean {
    return target1.targetValue === target2.targetValue
  }

  serialize(target: Target): string {
    return target.targetValue.toString()
  }

  deserialize(data: string): Target {
    if (typeof data !== 'string') {
      throw new Error('Can only deserialize Target from string')
    }
    return new Target(data)
  }
}

export const TargetSerdeInstance = new TargetSerde()
