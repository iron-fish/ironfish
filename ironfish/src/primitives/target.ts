/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

import { BigIntUtils } from '../utils/bigint'

/**
 *  Minimum difficulty, which is equivalent to maximum target
 */
const MIN_DIFFICULTY = 131072n

/**
 * Maximum target, which is equivalent of minimum difficulty of 131072
 * target == 2**256 / difficulty
 */
const MAX_TARGET = 883423532389192164791648750371459257913741948437809479060803100646309888n

/**
 *  Maximum number to represent a 256 bit number, which is 2**256 - 1
 */
const MAX_256_BIT_NUM =
  115792089237316195423570985008687907853269984665640564039457584007913129639935n

export class Target {
  targetValue: bigint
  constructor(targetValue: bigint | Buffer | string | number) {
    const candidate =
      targetValue instanceof Buffer ? BigIntUtils.fromBytesBE(targetValue) : BigInt(targetValue)

    if (candidate > MAX_256_BIT_NUM) {
      throw new Error('Target value exceeds max target')
    } else {
      this.targetValue = candidate
    }
  }

  /**
   * Maximum target (in terms of difficulty), which is equivalent of
   * minimum difficulty of 131072
   * maximum target == minimum difficulty
   * target == 2**256 / difficulty
   */
  static maxTarget(): Target {
    return new Target(MAX_TARGET)
  }

  static minTarget(): Target {
    return new Target(0)
  }

  /**
   * Calculate the target for the current block given the timestamp in that
   * block's header, the previous block's timestamp and previous block's target.
   *
   * To verify whether a target on a block is correct, pass in the timestamp in its header,
   * its previous block's timestamp, and its previous block's target
   * and compare the resulting target to what is specified on the current block header
   *
   * @param time the block's timestamp for which the target is calculated for
   * @param previousBlockTimestamp the block's previous block header's timestamp
   * @param previousBlockTarget the block's previous block header's target
   */
  static calculateTarget(
    time: Date,
    previousBlockTimestamp: Date,
    previousBlockTarget: Target,
    targetBlockTimeInSeconds: number,
    targetBucketTimeInSeconds: number,
  ): Target {
    const parentDifficulty = previousBlockTarget.toDifficulty()

    const difficulty = Target.calculateDifficulty(
      time,
      previousBlockTimestamp,
      parentDifficulty,
      targetBlockTimeInSeconds,
      targetBucketTimeInSeconds,
    )

    return Target.fromDifficulty(difficulty)
  }

  /**
   * Calculate the difficulty for the current block given the timestamp in that
   * block's header, the previous block's timestamp and previous block's target.
   *
   * Note that difficulty == 2**256 / target and target == 2**256 / difficulty
   *
   * Algorithm: difficulty = parentDifficulty - (parentDifficulty / 2048) * bucket
   * Where bucket is how many steps (in TARGET_BUCKET_TIME_IN_SECONDS) the new time is away from
   * our target bucket range, e.g. for target block time of 60 seconds (with +/-5 seconds forgiveness):
   * 35 - 45 seconds: bucket -2
   * 45 - 55 seconds: bucket -1
   * 55 - 65 seconds: bucket 0
   * 65 - 75 seconds: bucket 1
   * 75 - 85 seconds: bucket 2
   * .. and so on
   *
   * Returns the difficulty for a block given it timestamp for that block and its parent.
   * @param time the block's timestamp for which the target is calculated for
   * @param previousBlockTimestamp the block's previous block header's timestamp
   * @param previousBlockTarget the block's previous block header's target
   */
  static calculateDifficulty(
    time: Date,
    previousBlockTimestamp: Date,
    previousBlockDifficulty: bigint,
    targetBlockTimeInSeconds: number,
    targetBucketTimeInSeconds: number,
  ): bigint {
    const diffInSeconds = (time.getTime() - previousBlockTimestamp.getTime()) / 1000

    let bucket = Math.floor(
      (diffInSeconds - targetBlockTimeInSeconds + Math.floor(targetBucketTimeInSeconds / 2)) /
        targetBucketTimeInSeconds,
    )

    // Should not change difficulty by more than 99 buckets from last block's difficulty
    bucket = Math.min(bucket, 99)

    const difficulty =
      previousBlockDifficulty - (previousBlockDifficulty / 2048n) * BigInt(bucket)

    return BigIntUtils.max(difficulty, Target.minDifficulty())
  }

  /**
   * Returns the minimum difficulty that can be used for Iron Fish blocks
   * To be used in calculateTarget for easier mocking
   */
  static minDifficulty(): bigint {
    return MIN_DIFFICULTY
  }

  /**
   * Converts difficulty to Target
   */
  static fromDifficulty(difficulty: bigint): Target {
    if (difficulty <= Target.minDifficulty()) {
      return Target.maxTarget()
    }
    return new Target((2n ** 256n / difficulty).valueOf())
  }

  /**
   * Return the difficulty representation as a big integer
   */
  toDifficulty(): bigint {
    if (this.targetValue <= 1n) {
      return MAX_256_BIT_NUM
    }
    return 2n ** 256n / this.targetValue
  }

  /**
   * Add the given amount to the target's value. A negative amount makes the target
   * harder to achieve, a positive one makes it easier.
   *
   * If adjustment would make target negative or higher than max allowed value,
   * the current target is returned unchanged.
   */
  adjust(amount: bigint): Target {
    let adjusted = this.targetValue + amount
    if (adjusted > MAX_TARGET || adjusted < 0) {
      adjusted = this.targetValue
    }
    return new Target(adjusted)
  }

  /**
   * Return whether or not this target meets the requirements of the given target,
   * which is to say, this has a lower numeric value then the provided one.
   */
  static meets(hashValue: BigInt, target: Target): boolean {
    return hashValue <= target.targetValue
  }

  /**
   * Return the target number as a big integer
   */
  asBigInt(): bigint {
    return this.targetValue
  }

  equals(other: Target): boolean {
    return this.targetValue === other.targetValue
  }
}
