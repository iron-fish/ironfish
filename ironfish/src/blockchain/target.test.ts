/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

import { Target, TargetSerde, bigIntToBytes, bytesToBigInt } from './target'

describe('Target', () => {
  it('converts bigints to bytes and back', () => {
    const bigints = [
      BigInt(0),
      BigInt(
        '9999999999999999999999999999999999999999999999999999999999999999999999999999999999',
      ),
      BigInt(255),
      BigInt(256),
      BigInt(1024),
      BigInt(1025),
    ]
    for (const candidate of bigints) {
      const bytes = bigIntToBytes(candidate)
      const back = bytesToBigInt(bytes)
      expect(back).toEqual(candidate)
    }
  })

  it('converts empty array to 0', () => {
    expect(bytesToBigInt(Buffer.from([]))).toEqual(BigInt(0))
  })

  it('constructs targets', () => {
    expect(new Target().asBigInt()).toEqual(BigInt(0))
    expect(new Target(BigInt(9999999999999)).asBigInt()).toEqual(BigInt(9999999999999))
    expect(new Target(Buffer.from([4, 8])).asBigInt()).toEqual(BigInt('1032'))
    expect(new Target(Buffer.from([0, 0, 0, 0, 0, 0, 0, 4, 8])).asBigInt()).toEqual(
      BigInt('1032'),
    )
    expect(new Target(Buffer.alloc(32)).asBigInt()).toEqual(BigInt('0'))
  })

  it('makes the correct bytes', () => {
    const bigints = [
      BigInt(0),
      BigInt('99999999999999999999999999999999999999999999999999999999999999999999999'),
      BigInt(255),
      BigInt(256),
      BigInt(1024),
      BigInt(1025),
    ]
    for (const candidate of bigints) {
      expect(new Target(candidate).asBytes()).toMatchSnapshot()
    }
  })

  it('throws when constructed with too big an array', () => {
    const bytes = Buffer.alloc(33)
    bytes[0] = 1
    expect(() => new Target(bytes)).toThrowErrorMatchingInlineSnapshot(
      `"Target value exceeds max target"`,
    )
  })

  it('has the correct max value', () => {
    // The minimum difficulty is 131072, which means the maximum target is 2**256 / 131072
    const maxTarget = BigInt(2) ** BigInt(256) / BigInt(Target.minDifficulty())
    expect(Target.maxTarget().asBigInt()).toBe(maxTarget)
  })

  it('adjusts target up', () => {
    expect(new Target('55').adjust(BigInt('5')).targetValue).toEqual(BigInt('60'))
  })

  it('adjusts target down', () => {
    expect(new Target('55').adjust(BigInt('-5')).targetValue).toEqual(BigInt('50'))
  })

  it("doesn't adjust negative", () => {
    expect(new Target('55').adjust(BigInt('-60')).targetValue).toEqual(BigInt('55'))
  })

  it("doesn't adjust past max", () => {
    expect(Target.maxTarget().adjust(BigInt('-5')).adjust(BigInt('10')).targetValue).toEqual(
      Target.maxTarget().adjust(BigInt('-5')).targetValue,
    )
  })

  it('meets other target values', () => {
    const target = new Target('43')
    expect(Target.meets(BigInt(42), target)).toBe(true)
    expect(Target.meets(BigInt(43), target)).toBe(true)
    expect(Target.meets(BigInt(44), target)).toBe(false)
  })
})

describe('TargetSerde', () => {
  const serde = new TargetSerde()
  it('checks target equality', () => {
    expect(serde.equals(new Target('588888'), new Target('588888'))).toBe(true)
  })
  it('serializes and deserializes bytes', () => {
    const target = new Target(500)
    const serialized = serde.serialize(target)
    expect(serialized).toMatchInlineSnapshot(`"500"`)
    const deserialized = serde.deserialize(serialized)
    expect(serde.equals(deserialized, target)).toBe(true)
  })
  it('throws when deserializing incorrect value', () => {
    expect(() => serde.deserialize('not a number')).toThrowErrorMatchingInlineSnapshot(
      `"Cannot convert not a number to a BigInt"`,
    )
    // @ts-expect-error Argument of type '{ not: string; }' is not assignable to parameter of type 'string'.ts(2345)
    expect(() => serde.deserialize({ not: 'a string' })).toThrowErrorMatchingInlineSnapshot(
      `"Can only deserialize Target from string"`,
    )
  })
})

describe('Calculate target', () => {
  it('can increase target (which decreases difficulty) if its taking too long to mine a block (20+ seconds since last block)', () => {
    const now = new Date()
    // for any time 20-29 seconds after the last block, difficulty should decrease by previous block's difficulty / BigInt(2048)
    for (let i = 1; i < 10; i++) {
      const time = new Date(now.getTime() + 20000 + i * 1000)

      const difficulty = BigInt(231072)
      const target = Target.fromDifficulty(difficulty)

      const diffInDifficulty = difficulty / BigInt(2048)

      const newDifficulty = Target.calculateDifficulty(time, now, difficulty)
      const newTarget = Target.calculateTarget(time, now, target)

      expect(newDifficulty).toBeLessThan(difficulty)
      expect(BigInt(newDifficulty) + diffInDifficulty).toEqual(difficulty)

      expect(newTarget.asBigInt()).toBeGreaterThan(target.asBigInt())
    }

    // for any time 30-39 seconds after the last block, difficulty should decrease by previous block's difficulty / BigInt(2048) * 2
    for (let i = 1; i < 10; i++) {
      const time = new Date(now.getTime() + 30000 + i * 1000)

      const difficulty = BigInt(231072)
      const target = Target.fromDifficulty(difficulty)

      const diffInDifficulty = (difficulty / BigInt(2048)) * BigInt(2)

      const newDifficulty = Target.calculateDifficulty(time, now, difficulty)
      const newTarget = Target.calculateTarget(time, now, target)

      expect(newDifficulty).toBeLessThan(difficulty)
      expect(BigInt(newDifficulty) + diffInDifficulty).toEqual(difficulty)

      expect(newTarget.asBigInt()).toBeGreaterThan(target.asBigInt())
    }
  })

  it('can decrease target (which increases difficulty) if a block is trying to come in too early (1-10 seconds)', () => {
    const now = new Date()
    for (let i = 1; i < 10; i++) {
      const time = new Date(now.getTime() + i * 1000)

      const difficulty = BigInt(231072)
      const target = Target.fromDifficulty(difficulty)

      const diffInDifficulty = difficulty / BigInt(2048)

      const newDifficulty = Target.calculateDifficulty(time, now, difficulty)
      const newTarget = Target.calculateTarget(time, now, target)

      expect(newDifficulty).toBeGreaterThan(difficulty)
      expect(BigInt(difficulty) + diffInDifficulty).toEqual(newDifficulty)

      expect(newTarget.targetValue).toBeLessThan(target.targetValue)
    }
  })

  it('keeps difficulty/target of parent block header if time differnece is between 10 and 20 seconds', () => {
    const now = new Date()
    for (let i = 10; i < 20; i++) {
      const time = new Date(now.getTime() + i * 1000)

      const difficulty = BigInt(231072)
      const target = Target.fromDifficulty(difficulty)

      const newDifficulty = Target.calculateDifficulty(time, now, difficulty)
      const newTarget = Target.calculateTarget(time, now, target)

      const diffInDifficulty = BigInt(newDifficulty) - difficulty

      expect(diffInDifficulty).toEqual(BigInt(0))
      expect(newDifficulty).toEqual(difficulty)
      expect(newTarget.targetValue).toEqual(target.targetValue)
    }
  })
})
