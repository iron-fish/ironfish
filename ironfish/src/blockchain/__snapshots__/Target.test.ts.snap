// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Target makes the correct bytes 1`] = `
Object {
  "data": Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  "type": "Buffer",
}
`;

exports[`Target makes the correct bytes 2`] = `
Object {
  "data": Array [
    0,
    0,
    14,
    125,
    52,
    198,
    74,
    156,
    133,
    212,
    70,
    13,
    187,
    202,
    135,
    25,
    107,
    97,
    97,
    138,
    75,
    210,
    22,
    127,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
  ],
  "type": "Buffer",
}
`;

exports[`Target makes the correct bytes 3`] = `
Object {
  "data": Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    255,
  ],
  "type": "Buffer",
}
`;

exports[`Target makes the correct bytes 4`] = `
Object {
  "data": Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    1,
    0,
  ],
  "type": "Buffer",
}
`;

exports[`Target makes the correct bytes 5`] = `
Object {
  "data": Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    4,
    0,
  ],
  "type": "Buffer",
}
`;

exports[`Target makes the correct bytes 6`] = `
Object {
  "data": Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    4,
    1,
  ],
  "type": "Buffer",
}
`;
